packages/astro/src/core/render-context.ts

```diff
@@ -80,22 +91,51 @@ export class RenderContext {
 	 * - fallback
 	 */
 	async render(componentInstance: ComponentInstance | undefined): Promise<Response> {
-		const { cookies, middleware, pathname, pipeline, routeData } = this;
+		const { cookies, middleware, pathname, pipeline } = this;
 		const { logger, routeCache, serverLike, streaming } = pipeline;
 		const props = await getProps({
 			mod: componentInstance,
-			routeData,
+			routeData: this.routeData,
 			routeCache,
 			pathname,
 			logger,
 			serverLike,
 		});
 		const apiContext = this.createAPIContext(props);
 
-		const lastNext = async () => {
-			switch (routeData.type) {
+		this.counter++;
+		if (this.counter === 4) {
+			return new Response('Loop Detected', {
+				// https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/508
+				status: 508,
+				statusText: 'Astro detected a loop where you tried to call the rewriting logic more than four times.',
+			});
+		}
+		const lastNext = async (ctx: APIContext, payload?: RewritePayload) => {
+			if (payload) {
+				if (this.pipeline.manifest.rewritingEnabled) {
+					try {
+						const [routeData, component] = await pipeline.tryRewrite(payload);
+						this.routeData = routeData;
+						componentInstance = component;
+					} catch (e) {
+						return new Response('Not found', {
+							status: 404,
+							statusText: 'Not found',
+						});
+					} finally {
+						this.isRerouting = true;
+					}
+				} else {
+					this.pipeline.logger.warn(
+						'router',
+						'You tried to use the routing feature without enabling it via experimental flag. This is not allowed.'
```

Comment:
```suggestion
						'The rewrite API is experimental. To use this feature, add the `rewriting` flag to the `experimental` object in your Astro config.'
```

ditto on this warning